packages/ui/report.api.md

```diff
@@ -1535,48 +1781,69 @@ export const useBreakpoint: () => {
   down: (key: Breakpoint) => boolean;
 };
 
-// @public
-export function useTable<T = any>(
-  config?: UseTableConfig<T>,
-): UseTableResult<T>;
+// @public (undocumented)
+export function useTable<T extends TableItem, TFilter = unknown>(
+  options: UseTableOptions<T, TFilter>,
+): UseTableResult<T, TFilter>;
 
 // @public (undocumented)
-export interface UseTableConfig<T = any> {
-  data?: T[];
-  pagination?: UseTablePaginationConfig;
+export interface UseTableCompleteOptions<T extends TableItem, TFilter = unknown>
+  extends QueryOptions<TFilter> {
+  // (undocumented)
+  filterFn?: (data: T[], filter: TFilter) => T[];
+  // (undocumented)
+  getData: () => T[] | Promise<T[]>;
+  // (undocumented)
+  mode: 'complete';
+  // (undocumented)
+  paginationOptions?: PaginationOptions;
+  // (undocumented)
+  searchFn?: (data: T[], search: string) => T[];
+  // (undocumented)
+  sortFn?: (data: T[], sort: SortDescriptor) => T[];
 }
 
 // @public (undocumented)
-export interface UseTablePagination<T = any> {
-  data?: T[];
-  nextPage: () => void;
-  offset: number;
-  pageSize: number;
-  paginationProps: TablePaginationProps;
-  previousPage: () => void;
-  setOffset: (offset: number) => void;
-  setPageSize: (pageSize: number) => void;
+export interface UseTableCursorOptions<T extends TableItem, TFilter = unknown>
+  extends QueryOptions<TFilter> {
+  // (undocumented)
+  getData: (params: CursorParams<TFilter>) => Promise<CursorResponse<T>>;
+  // (undocumented)
+  mode: 'cursor';
+  // (undocumented)
+  paginationOptions?: Omit<PaginationOptions, 'initialOffset'>;
 }
 
 // @public (undocumented)
-export interface UseTablePaginationConfig {
-  defaultOffset?: number;
-  defaultPageSize?: number;
-  offset?: number;
-  onNextPage?: () => void;
-  onOffsetChange?: (offset: number) => void;
-  onPageSizeChange?: (pageSize: number) => void;
-  onPreviousPage?: () => void;
-  pageSize?: number;
-  rowCount?: number;
-  showPageSizeOptions?: boolean;
+export interface UseTableOffsetOptions<T extends TableItem, TFilter = unknown>
+  extends QueryOptions<TFilter> {
+  // (undocumented)
+  getData: (params: OffsetParams<TFilter>) => Promise<OffsetResponse<T>>;
+  // (undocumented)
+  mode: 'offset';
+  // (undocumented)
+  paginationOptions?: PaginationOptions;
 }
 
 // @public (undocumented)
-export interface UseTableResult<T = any> {
-  data?: T[];
-  pagination: UseTablePagination<T>;
-  paginationProps: TablePaginationProps;
+export type UseTableOptions<T extends TableItem, TFilter = unknown> =
+  | UseTableCompleteOptions<T, TFilter>
+  | UseTableOffsetOptions<T, TFilter>
+  | UseTableCursorOptions<T, TFilter>;
```

Comment:
Worth checking whether overloads of `useTable` provides better TS completions compared to this union

Reply:
The union gives full completion on the options (e.g. you only see `sortFn` if you set mode to `complete`), so I think we're covered. Or did you mean something else?